### PR TITLE
ci: run different jobs depending on the modified files

### LINKED ISSUE #2233: Proposal: run different CI depending on the modified files
## Feature Request

**Is your feature request related to a problem? Please describe:**

Yes, this proposal relates to the long-standing question of why our CI has to be run every time, even if just one `.md` file is modified.

**Describe the feature you'd like:**

Run tests based on the type of change files.

**Teachability, Documentation, Adoption, Migration Strategy:**

https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-including-paths

### PR DESCRIPTION
**Pull request by g1eny0ung**
Signed-off-by: Yue Yang <g1enyy0ung@gmail.com>

### What problem does this PR solve?

Close #2233.

### What's changed and how it works?

In addition to what the title says, this PR adds UI tests in GitHub Actions and also removes some tests because we decided to use an `OpenAPI` converter to generate HTTP APIs.

### Related changes

- [ ] Need to update `chaos-mesh/website`
- [ ] Need to update `Dashboard UI`
- Need to **cheery-pick to release branches**
  - [x] release-2.1
  - [ ] release-2.0

### Checklist

Tests

- [x] Unit test
- [ ] E2E test
- [ ] No code
- [ ] Manual test (add steps below)

Side effects

- [ ] Breaking backward compatibility

### Release note 

```text
Please add a release note.

You can safely ignore this section if you don't think this PR needs a release note.
```

### DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```

### CODE REVIEW

**Comment by g1eny0ung** [comment]
/cc @chaos-mesh/maintainers @chaos-mesh/committers

Here are the more detailed updates:

- The following workflows remove the `on push` event in the `master` branch (since we have already checked it at `pull request`):
  - [x] `checklink.yaml`
  - [x] `codecov_unittest.yaml`
  - [x] `integration_test.yml`
  - [x] `license_checker.yml`
  - [x] `script_test.yml`

- For a long time, we have always run several **required checks** before merging a PR into the master, and if we want to run different jobs depending on modified files, it's needed to skip unmatched jobs. But GitHub Action can't do this at the job level. So in `ci.yml`, I use [dorny/paths-filter@v2](https://github.com/dorny/paths-filter) to do this:
  - Example: https://github.com/dorny/paths-filter#conditional-execution.
  - **Note:** we need to update the settings of `Require status checks to pass before merging` to make modified workflows finally work.

**Comment by STRRL** [comment]
I am afraid that we could not remove the check with codecov_unittest.yaml, because we use squash commit when mething PRs, codecov would not find the correct commit as the base when diff the coverage with PR.

So we should upload the latest coverage report on `push`

**Comment by g1eny0ung** [comment]
> I am afraid that we could not remove the check with codecov_unittest.yaml, because we use squash commit when mething PRs, codecov would not find the correct commit as the base when diff the coverage with PR.
> 
> So we should upload the latest coverage report on `push`

Uh-oh, I forgot to think about it...I will re-add the push event in coverage workflow.

**Comment by g1eny0ung** [comment]
@STRRL PTAL again.

File: .github/workflows/ci.yml
Comment (STRRL) [inline .github/workflows/ci.yml:31]:
```suggestion
              - Makefile
```

I found that `paths-filter` uses `picomatch` for matching files, and `picomatch` follows case-sensitive by default.

File: .github/workflows/codecov_unittest.yaml
Comment (STRRL) [inline .github/workflows/codecov_unittest.yaml:7]: ditto

File: .github/workflows/script_test.yml
Comment (STRRL) [inline .github/workflows/script_test.yml:7]: I am not sure do we need to test the `install.sh` not only on `install.sh` changes, but also like changes in Dockerfile and CRD.

PTAL @YangKeao @Hexilee @iguoyr

**Review by STRRL — APPROVED** [review]
rest LGTM

File: .github/workflows/codecov_unittest.yaml
Comment (STRRL) [inline .github/workflows/codecov_unittest.yaml:12]: ditto, use `Makefile`

**Comment by STRRL** [comment]
sorry, I just want to click "Comment", not "Approve". Please take a look on the comments @g1eny0ung

**Comment by g1eny0ung** [comment]
> sorry, I just want to click "Comment", not "Approve". Please take a look on the comments @g1eny0ung

Got it! I'm have already fixed them.

File: .github/workflows/script_test.yml
Comment (iguoyr) [reply .github/workflows/script_test.yml:7]: Integration test will run `hack/local-up-chaos-mesh.sh`, and it calls `install.sh`, will it solve this problem?

**Review by STRRL — APPROVED** [review]

**Review by iguoyr — APPROVED** [review]
LGTM

**Pull request merged**
